README.md
# Crypto Multihash

[![Build Status](https://travis-ci.org/mseri/crypto-multihash.svg?branch=master)](https://travis-ci.org/mseri/crypto-multihash)
[![Hackage](https://img.shields.io/hackage/v/crypto-multihash.svg)](http://hackage.haskell.org/package/crypto-multihash)
![Hackage Dependencies](https://img.shields.io/hackage-deps/v/crypto-multihash.svg)
![Haskell Programming Language](https://img.shields.io/badge/language-Haskell-blue.svg)
![BSD3 License](http://img.shields.io/badge/license-BSD3-brightgreen.svg)


Multihash library implemented on top of [cryptonite](https://hackage.haskell.org/package/cryptonite) cryptographic library. 
Multihash is a protocol for encoding the hash algorithm and digest length at the start of the digest, see the official [multihash github page](https://github.com/jbenet/multihash/).

This library is still experimental and the api is not guaranteed stable. 
I will increment the version number appropriately in case of breaking changes.

For the moment the library implements all the expected hashing algorithms with the exception of shake-128 and shake-256. A Multihash can be encoded in hex (`Base16`), bitcoin base58 (`Base58`) and base64 (`Base64`). 

The `Base32` encoding is not yet supported due to discrepancy between the encoding from `Data.ByteArray.Encoding` and the one appearing in the official multihash page.

# Usage

```{.haskell}
-- in ghci `:set -XOverloadedStrings`
{-# LANGUAGE OverloadedStrings #-}

-- `:m +Crypto.Multihash`
import Crypto.Multihash
import Data.ByteString (ByteString)

main = do
    let v = "test"::ByteString
    let m = multihash SHA256 v

    -- If using the Weak module
    -- let m' = weakMultihash "sha256" v
    
    putStrLn $ "Base16: " ++ (encode' Base16 m)
    -- You might need to specify the encoded string type
    putStrLn $ "Base58: " ++ (encode' Base58 m :: String)

    -- `encode` is the safe interface returning an `Either` type
    putStrLn $ "Base64: " ++ show (encode Base64 m :: Either String String)
    
    let h = encode' Base58 m :: ByteString
    -- You can check that a multihash corresponds to some data `v`
    checkMultihash h v
    -- Right True
    
    -- Or if you have a Multihash to compare you can use it
    check h m
    -- Right True

    -- There is also an unsafe version, as for encode
    -- note that sometimes you will need to specify the string types
    checkMultihash' ("whatever"::String) v
    -- *** Exception: Unable to infer an encoding
    checkMultihash' ("Eiwhatever"::ByteString) v
    -- *** Exception: base64: input: invalid length
    check' ("EiCfhtCBiEx9ZZov6qDFWtAVo79PGysLgizRXWwVsPA1CA=="::ByteString) m
    -- False

    checkMultihash' h v
    -- True
    check' h m
    -- True
```

The of `import Crypto.Multihash.Weak` is almost identical, but it additionally introduces the function `toWeakMultihash` that tries to import a string as a `WeakMultihashDigest`.

# Test

Some preliminary tests can be performed with `stack test`. 

A simple example encoder is in `app/Main.hs`. 
You can run it on files

```{.bash}
echo -n test | stack exec mh -- somefile someotherfile
```

or read data from the standard input 

```{.bash}
echo -n test | stack exec mh -- -
```

# Contribution

1. Fork repository
2. Do some changes
3. Create pull request
4. Wait for CI build and review

You can use stack to build the project: `stack build`

To run tests: `stack test`

# TODO

- Test the new `getBase` implementation using quickcheck (critical, I will not push new version until this is done)
- Test the correct support of truncated multihashes
- Implement benchmarks, then start optimising the code where possible
- ~~Use the hash length in `checkPayload` to treat correctly truncated hashes (see https://github.com/jbenet/multihash/issues/1#issuecomment-91783612)~~
- Improve documentation
- Improve testing for for raised exceptions
- Implement `shake-128` and `shake-256` multihashes
- Implement `Base32` encoding waiting for https://github.com/jbenet/multihash/issues/31 to be resolved)


CHANGELOG.md
# CHANGELOG

- v0.4.0.0: major first API refactoring with the introduction of more generic interface and a new ``Weakly-typed'' API

test/Spec.hs
{-# LANGUAGE OverloadedStrings #-}

import Control.Monad (zipWithM)
import Crypto.Multihash
import Crypto.Multihash.Weak (weakMultihash, checkWeakMultihash, toWeakMultihash)
import Data.ByteString (ByteString, pack, unpack)
import Text.Printf (printf)

import Test.Hspec
import Test.QuickCheck

-- TODO: 
--   * use QuickCheck to test Multihash serializatio/deserializatio/check properties
--     especially now that we infer the decoding for arbitrary truncations
--   * test for valid and invalid truncated multihashes

instance Arbitrary ByteString where
    arbitrary = pack `fmap` arbitrary
    --coarbitrary = coarbitrary . unpack

testString :: ByteString
testString = "test"

failTestString :: ByteString
failTestString = "test1"

weakAlgos = [ "sha1", "sha256", "sha512", "sha3-512" 
            , "sha3-384", "sha3-256", "sha3-224"
            , "blake2b-512", "blake2s-256" ]

-- Useful for testing, generated in Weak by
-- tester = map (\(a,_) -> _getLength $ weakMultihash' a ("test"::ByteString)) allowedAlgos
maxHashLengths = [20,32,64,64,48,32,28,64,32]

main :: IO ()
main = runSpec

prop_gen_check :: ByteString -> Property
prop_gen_check str = 
  let m = multihash SHA1 str
      enc :: ByteString
      enc = encode' Base16 m 
  in check' enc m === True

runQC = quickCheck prop_gen_check

runSpec = hspec $ do
  describe "Multihash: check properties with QuickCheck" $
    it "correctly encodes and checks SHA1" $
      property prop_gen_check

  mhEncoding SHA1 h1
  mhCheck mh checkMultihash SHA1 h1
  mhEncoding SHA256 h2
  mhCheck mh checkMultihash SHA256 h2
  mhEncoding SHA512 h3
  mhCheck mh checkMultihash SHA512 h3
  mhEncoding SHA3_512 h4
  mhCheck mh checkMultihash SHA3_512 h4
  mhEncoding SHA3_384 h5
  mhCheck mh checkMultihash SHA3_384 h5
  mhEncoding SHA3_256 h6
  mhCheck mh checkMultihash SHA3_256 h6
  mhEncoding SHA3_224 h7
  mhCheck mh checkMultihash SHA3_224 h7
  mhEncoding Blake2b_512 h8
  mhCheck mh checkMultihash Blake2b_512 h8
  mhEncoding Blake2s_256 h9
  mhCheck mh checkMultihash Blake2s_256 h9
  ------------------------------------------------------------------------
  traverse (uncurry wmhEncoding) 
           (zip weakAlgos h)
  traverse (uncurry (mhCheck wmh checkWeakMultihash))
           (zip weakAlgos h)
  ------------------------------------------------------------------------
  describe "Multihash: fails correctly when" $ do
    it "checking an invalid truncated multihash" $
      checkMultihash ("1340ee26b0dd4af7e749aa1a8e"::ByteString) testString 
        `shouldBe` Left "Corrupted MultihasDigest: invalid length"
    it "checking an invalid truncated multihash" $
      checkMultihash ("dd4af7e749aa1a8e1340ee26b0"::ByteString) testString 
        `shouldBe` Left "Corrupted MultihasDigest: invalid length"

  describe "Weak Multihash: fails correctly when" $ do
    it "checking an invalid truncated multihash" $
      checkWeakMultihash ("1340ee26b0dd4af7e749aa1a8e"::ByteString) testString 
        `shouldBe` Left "Corrupted MultihasDigest: invalid length"
    it "checking an invalid truncated multihash" $
      checkWeakMultihash ("dd4af7e749aa1a8e1340ee26b0"::ByteString) testString 
        `shouldBe` Left "Corrupted MultihasDigest: invalid length"
  ------------------------------------------------------------------------

  where
    mh = "Multihash"::String
    wmh = "Weak Multihash"::String

    mhEncoding :: (HashAlgorithm a, Codable a, Show a) => a 
                      -> (ByteString, ByteString, ByteString) -> SpecWith ()
    mhEncoding alg (sm16, sm58, sm64) = 
        let m = multihash alg testString in
        describe (printf "Multihash: encoding %s multihash" (show alg)) $ do
          it "returns the correct Base16 hash" $ 
            encode' Base16 m `shouldBe` sm16
          it "returns the correct Base58 hash" $ 
            encode' Base58 m `shouldBe` sm58
          it "returns the correct Base64 hash" $ 
            encode' Base64 m `shouldBe` sm64

    wmhEncoding :: ByteString 
                       -> (ByteString, ByteString, ByteString) -> SpecWith ()
    wmhEncoding alg (sm16, sm58, sm64) = 
        let m = case weakMultihash alg testString of 
                  Right val -> val
                  Left  err -> error err
        in do
        describe (printf "Weak Multihash: encoding %s multihash" (show alg)) $ do
          it "returns the correct Base16 hash" $ 
            encode' Base16 m `shouldBe` sm16
          it "returns the correct Base58 hash" $ 
            encode' Base58 m `shouldBe` sm58
          it "returns the correct Base64 hash" $ 
            encode' Base64 m `shouldBe` sm64

        describe (printf "Weak Multihash: decoding %s multihash" (show alg)) $ do
          it "imports the correct hash from Base16" $ 
            toWeakMultihash sm16 `shouldBe` Right m
          it "imports the correct hash from Base58" $ 
            toWeakMultihash sm58 `shouldBe` Right m
          it "imports the correct hash from Base64" $ 
            toWeakMultihash sm64 `shouldBe` Right m
    
    -- hashChecker :: (HashAlgorithm a, Codable a, Show a) => a 
    --                -> (ByteString, ByteString, ByteString) -> SpecWith ()
    mhCheck t checker alg (e16, e58, e64) =
      describe (printf "%s: using checkPayload on %s hashes" t (show alg)) $ do
        it "checks correctly Base16 hashes"  $
          checker e16 testString `shouldBe` Right True
        it "fails correctly on Base16 hashes" $
          checker e16 failTestString `shouldBe` Right False
        it "checks correctly Base58 hashes" $
          checker e58 testString `shouldBe` Right True
        it "fails correctly on Base58 hashes" $
          checker e58 failTestString `shouldBe` Right False
        it "checks correctly Base64 hashes" $
          checker e64 testString `shouldBe` Right True
        it "fails correctly on Base64 hashes" $
          checker e64 failTestString `shouldBe` Right False

    -- array of triples of hashes of the string "test"
    h@[h1, h2, h3, h4, h5, h6, h7, h8, h9] = 
      [
        ( "1114a94a8fe5ccb19ba61c4c0873d391e987982fbbd3"
        , "5dt9CqvXK9qs7vazf7k7ZRqe28VPTg"
        , "ERSpSo/lzLGbphxMCHPTkemHmC+70w==")
      , ( "12209f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08"
        , "QmZ5NmGeStdit7tV6gdak1F8FyZhPsfA843YS9f2ywKH6w"
        , "EiCfhtCBiEx9ZZov6qDFWtAVo79PGysLgizRXWwVsPAKCA==")
      , ( "1340ee26b0dd4af7e749aa1a8ee3c10ae9923f618980772e473f8819a5d4940e0db27ac185f8a0e1d5f84f88bc887fd67b143732c304cc5fa9ad8e6f57f50028a8ff"
        , "8VxYhLL7s5BvLogtUGgNJ7DebZ5Ba9mG3izfc7v6o4RZ28x469vJaifa3TQ13Z9DycmAuJWnp7ErkZofm4rsMo78fQ"
        , "E0DuJrDdSvfnSaoajuPBCumSP2GJgHcuRz+IGaXUlA4NsnrBhfig4dX4T4i8iH/WexQ3MsMEzF+prY5vV/UAKKj/")
      , ( "14409ece086e9bac491fac5c1d1046ca11d737b92a2b2ebd93f005d7b710110c0a678288166e7fbe796883a4f2e9b3ca9f484f521d0ce464345cc1aec96779149c14"
        , "8tXEcJyq2MCx27UHYbZxmte37ezBawV35QhfKPtq5QeSnX66q4DDf1cwMYUh2pUVbxdQgrDaSjbrPrfNxzvSSLQAtT"
        , "FECezghum6xJH6xcHRBGyhHXN7kqKy69k/AF17cQEQwKZ4KIFm5/vnlog6Ty6bPKn0hPUh0M5GQ0XMGuyWd5FJwU")
      , ( "1530e516dabb23b6e30026863543282780a3ae0dccf05551cf0295178d7ff0f1b41eecb9db3ff219007c4e097260d58621bd"
        , "G9LerVN7c3uUAAymoAGkCGZPn53PZi1SHwPJ2nznLp82jcM2M1KLwpfrZh3F1QRVG3f2"
        , "FTDlFtq7I7bjACaGNUMoJ4Cjrg3M8FVRzwKVF41/8PG0Huy52z/yGQB8TglyYNWGIb0=")
      , ( "162036f028580bb02cc8272a9a020f4200e346e276ae664e45ee80745574e2f5ab80"
        , "W1d9SeHn1mCnY3jZMs5YeqfFbwEnq5gQy1VDymGoPK28RD"
        , "FiA28ChYC7AsyCcqmgIPQgDjRuJ2rmZORe6AdFV04vWrgA==")
      , ( "171c3797bf0afbbfca4a7bbba7602a2b552746876517a7f9b7ce2db0ae7b"
        , "5daZNVMeTfSuCvu7rBKsFkzEMebnuGjNpos1ThF1c"
        , "Fxw3l78K+7/KSnu7p2AqK1UnRodlF6f5t84tsK57")
      , ( "4040a71079d42853dea26e453004338670a53814b78137ffbed07603a41d76a483aa9bc33b582f77d30a65e6f29a896c0411f38312e1d66e0bf16386c86a89bea572"
        , "S2XUqUDxz3MHMZtJpCZKt5oRjXHQ34gsyDBT759qNwoSP9rDBHVHxjQUQtXfExotxTqf4rMEXQkNmXE3N9mhoZX6wK"
        , "QECnEHnUKFPeom5FMAQzhnClOBS3gTf/vtB2A6QddqSDqpvDO1gvd9MKZebymolsBBHzgxLh1m4L8WOGyGqJvqVy")
      , ( "4120f308fc02ce9172ad02a7d75800ecfc027109bc67987ea32aba9b8dcc7b10150e"
        , "2UPuEK7FVakwP3yUak5jKQhZb6pgpbcqYoRZ2tDzgeCfVr5"
        , "QSDzCPwCzpFyrQKn11gA7PwCcQm8Z5h+oyq6m43MexAVDg==")
      ]



app/Main.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Crypto.Multihash
import Data.ByteString (ByteString)
import qualified Data.ByteString as B
import Data.List
import System.Console.GetOpt
import System.IO hiding (withFile)
import System.Environment
import System.Exit
import Text.Printf (printf)

main :: IO ()
main = do
  (_args, files) <- getArgs >>= parse
  mapM_ printers files
  putStrLn "Done! Note: shake-128/256 and Base32 are not yet part of the library"
 
printer :: (HashAlgorithm a, Codable a, Show a) => a -> ByteString -> IO ()
printer alg bs = do
  let m = multihash alg bs
  putStrLn $ printf "Hash algorithm: %s" (show alg)
  putStrLn $ printf "Base16: %s" (encode' Base16 m :: String)
  -- Base32 missing
  putStrLn $ printf "Base58: %s" (encode' Base58 m :: String)
  putStrLn $ printf "Base64: %s" (encode' Base64 m :: String)
  putStrLn ""

printers :: FilePath -> IO ()
printers f = do
  d <- withFile f
  putStrLn $ printf "Hashing %s\n" (if f == "-" then show d else show f)
  printer SHA1 d
  printer SHA256 d
  printer SHA512 d
  printer SHA3_512 d
  printer SHA3_384 d
  printer SHA3_256 d
  printer SHA3_224 d
  printer Blake2b_512 d
  printer Blake2s_256 d
  putStrLn ""
  where withFile f = if f == "-" then B.getContents else B.readFile f

data Flag = Help                  -- --help
          deriving (Eq,Ord,Enum,Show,Bounded)
 
flags :: [OptDescr Flag]
flags = [Option [] ["help"] (NoArg Help) "Print this help message"]

parse :: [String] -> IO ([Flag], [String])
parse argv = case getOpt Permute flags argv of
    (args,fs,[]) -> do
        let files = if null fs then ["-"] else fs
        if Help `elem` args
            then do hPutStrLn stderr (usageInfo header flags)
                    exitSuccess
            else return (nub (concatMap set args), files)
 
    (_,_,errs)   -> do
        hPutStrLn stderr (concat errs ++ usageInfo header flags)
        exitWith (ExitFailure 1)
 
    where header = "Usage: mh [file ...]"
          set f  = [f]


src/Crypto/Multihash/Internal.hs
-- |
-- Module      : Crypto.Multihash.Internal
-- License     : BSD3
-- Maintainer  : Marcello Seri <marcello.seri@gmail.com>
-- Stability   : experimental
-- Portability : unknown
--
{-# LANGUAGE OverloadedStrings #-}
module Crypto.Multihash.Internal where

import Data.ByteArray (ByteArrayAccess(..))
import qualified Data.ByteArray as BA
import qualified Data.ByteArray.Encoding as BE
import qualified Data.ByteString as BS
import qualified Data.ByteString.Base58 as B58
import Data.List (elemIndices)
import Data.Word (Word8)

-------------------------------------------------------------------------------
import Crypto.Multihash.Internal.Types
-------------------------------------------------------------------------------

-- | Convert a maybe type to an either type
maybeToEither :: l -> Maybe r -> Either l r
maybeToEither _ (Just res) = Right res
maybeToEither err _        = Left err

-- | Codes corresponding to the various hash algorithms
hashCodes :: [Word8]
hashCodes = map fromIntegral
                ([0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x40, 0x41]::[Int])

-- | Convert a 'BS.ByteString' from a 'Base' into a 'BS.ByteString' in 'Base2'.
convertFromBase :: Base -> BS.ByteString -> Either String BS.ByteString
convertFromBase b bs = case b of
  Base2  -> Left "This is not supposed to happen"
  Base16 -> BE.convertFromBase BE.Base16 bs
  Base32 -> Left "Base32 decoder not implemented"
  Base58 -> do
    dec <- maybeToEither "Base58 decoding error" (B58.decodeBase58 B58.bitcoinAlphabet bs)
    return (BA.convert dec)
  Base64 -> BE.convertFromBase BE.Base64 bs

-- | Infer the 'Base' encoding function from an encoded 'BS.BinaryString'.
--   Supports only 'Base16', bitcoin 'Base58' and 'Base64' for the moment.
--   NOTE: it can fail or _infer the wrong encoding_ if the string is too short
--   or if it is encoded using a different standard.
getBase :: BS.ByteString -> Either String Base
getBase h = if len == 0 
  then 
    Left "Unable to infer an encoding" 
  else 
    pure $ [Base16, Base58, Base64] !! (head bsi)
  where
    len = Prelude.length bsi
    bsi = elemIndices 0 $ map (unmatch h) [b16Alphabet, b58Alphabet, b64Alphabet]
    unmatch str alphabet = BS.length $ BS.filter (`BS.notElem` alphabet) str

    b16Alphabet :: BS.ByteString
    b16Alphabet = "0123456789abcdef"

    b58Alphabet :: BS.ByteString
    b58Alphabet = B58.unAlphabet B58.bitcoinAlphabet

    b64Alphabet :: BS.ByteString
    b64Alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="

-- TODO: test this with quickcheck

-- old implementation
-- getBase :: BS.ByteString -> Either String Base
-- getBase h
--       | startWiths h ["1114", "1220", "1340", "1440", "1530", "1620", "171c", "4040", "4120"] = Right Base16
--       | startWiths h ["5d", "Qm", "8V", "8t", "G9", "W1", "5d", "S2", "2U"] = Right Base58
--       | startWiths h ["ER", "Ei", "E0", "FE", "FT", "Fi", "Fx", "QE", "QS"] = Right Base64
--       | otherwise = Left "Unable to infer an encoding"
--       where startWiths h = any (`BS.isPrefixOf` h)


-- | Encode the binary data using a given 'Base'.
encoder :: ByteArrayAccess a => Base -> a -> Either String BA.Bytes
encoder base bs = case base of
            Base2  -> return $ BA.convert bs
            Base16 -> return $ BE.convertToBase BE.Base16 bs
            Base32 -> Left "Base32 encoder not implemented"
            Base58 -> return $ BA.convert $ B58.encodeBase58 B58.bitcoinAlphabet 
                                                             (BA.convert bs :: BS.ByteString)
            Base64 -> return $ BE.convertToBase BE.Base64 bs

-- | Compare the lenght of the encoded multihash digest with the encoded hash length.
--   Return 'True' if the lengths are matching.
badLength :: ByteArrayAccess bs => bs -> Bool
badLength mh = 
      case BA.length mh of
        n | n <= 2 -> True
        n | BA.index mh 1 /= (fromIntegral n-2) -> True
        _ -> False


src/Crypto/Multihash/Weak.hs
-- |
-- Module      : Crypto.Multihash.Weak
-- License     : BSD3
-- Maintainer  : Marcello Seri <marcello.seri@gmail.com>
-- Stability   : experimental
-- Portability : unknown
--
-- Multihash library built on top of haskell 'cryptonite' crypto package
-- Multihash is a protocol for encoding the hash algorithm and digest length 
-- at the start of the digest, see the official 
-- <https://github.com/jbenet/multihash/ multihash poroposal github repo>.
--
-- The 'Crypto.Multihash.Weak' module contains a weaker implementation of the
-- MultihashDigest type that allows to deserialize multihashes and to work with
-- different algorithms at the same type. The API interface is almost identical
-- to the one exposed by 'Crypto.Multihash'. The notable differences are:
-- 
--  - the hash algorithms is no longer a type 
--  - the type returned by the multihash constructors is an 'Either' 
--    (see 'weakMultihash' and 'weakMultihashlazy')
--  - there is a function 'toWeakMultihash' to deserialise multihashes
--
-- For additional informations refer to the README.md or the
-- <https://github.com/mseri/crypto-multihash gihub repository>.
--
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE FlexibleContexts #-}

-- TODO: use length in check* to treat correctly truncated hashes
-- see https://github.com/jbenet/multihash/issues/1#issuecomment-91783612

module Crypto.Multihash.Weak 
  ( -- * Weak Multihash Types
    WeakMultihashDigest
  , Base            (..)
  , Encodable       (..)
  --, Checkable       (..)
  --, Payload         (..)
    -- * Weak Multihash Helpers
  , weakMultihash
  , weakMultihash'
  , weakMultihashlazy
  , weakMultihashlazy'
  , truncatedWeakMultihash
  , truncatedWeakMultihash'
  , toWeakMultihash
  , checkWeakMultihash
  , checkWeakMultihash'
  , getBase
  ) where

import Crypto.Hash (Digest, hashWith, hashlazy)
import qualified Crypto.Hash.Algorithms as A
import Data.ByteArray (ByteArrayAccess, Bytes)
import qualified Data.ByteArray as BA
import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import qualified Data.ByteString.Lazy as BL
import Data.List (elemIndex)
import Data.String (IsString(..))
import Data.String.Conversions
import Data.Word (Word8)

-------------------------------------------------------------------------------
import Crypto.Multihash.Internal.Types
import Crypto.Multihash.Internal
-------------------------------------------------------------------------------

data HashAlgo = S1 A.SHA1 | S256 A.SHA256 | S512 A.SHA512 
              | S3_512 A.SHA3_512 | S3_384 A.SHA3_384
              | S3_256 A.SHA3_256 | S3_224 A.SHA3_224
              | B2s A.Blake2s_256 | B2b A.Blake2b_512

instance Eq HashAlgo where
  (S1 _)     == (S1 _)     = True 
  (S256 _)   == (S256 _)   = True
  (S512 _)   == (S512 _)   = True
  (S3_512 _) == (S3_512 _) = True
  (S3_384 _) == (S3_384 _) = True
  (S3_256 _) == (S3_256 _) = True
  (S3_224 _) == (S3_224 _) = True
  (B2s _)    == (B2s _)    = True
  (B2b _)    == (B2b _)    = True
  _          == _          = False

instance Show HashAlgo where
  show (S1 A.SHA1)         = "sha1"
  show (S256 A.SHA256)     = "sha256"
  show (S512 A.SHA512)     = "sha512"
  show (S3_512 A.SHA3_512) = "sha3-512"
  show (S3_384 A.SHA3_384) = "sha3-384"
  show (S3_256 A.SHA3_256) = "sha3-256"
  show (S3_224 A.SHA3_224) = "sha3-224"
  show (B2s A.Blake2s_256) = "blake2s-256"
  show (B2b A.Blake2b_512) = "blake2b-512"

instance Codable HashAlgo where
  toCode (S1 A.SHA1)         = 0x11
  toCode (S256 A.SHA256)     = 0x12
  toCode (S512 A.SHA512)     = 0x13
  toCode (S3_512 A.SHA3_512) = 0x14
  toCode (S3_384 A.SHA3_384) = 0x15
  toCode (S3_256 A.SHA3_256) = 0x16
  toCode (S3_224 A.SHA3_224) = 0x17
  toCode (B2b A.Blake2b_512) = 0x40
  toCode (B2s A.Blake2s_256) = 0x41

-------------------------------------------------------------------------------

-- | Weak Multihash Digest container
data WeakMultihashDigest = WeakMultihashDigest
  { _getAlgorithm :: HashAlgo      -- ^ hash algorithm encoded as int
  , _getLength    :: Int           -- ^ hash lenght
  , _getDigest    :: Bytes         -- ^ binary digest data
  } deriving (Eq)

instance Show WeakMultihashDigest where
  -- an error here should never happen
  show = encode' Base58

instance Encodable WeakMultihashDigest where
  encode base (WeakMultihashDigest alg len md) = 
    if len == BA.length md
      then do
        d <- fullDigestUnpacked
        return $ fromString $ map (toEnum . fromIntegral) d
      else 
        Left "Corrupted MultihashDigest: invalid length"
    where
      fullDigestUnpacked :: Either String [Word8]
      fullDigestUnpacked = do
        d <- encoder base fullDigest
        return $ BA.unpack d
      
      fullDigest :: Bytes
      fullDigest = BA.pack hd `BA.append` md
        where
          hd :: [Word8]
          hd = fromIntegral <$> [toCode alg, len]


  check hash_ multihash_ = let hash_' = convertString hash_ in do
    base <- getBase hash_'
    m <- encode base multihash_
    return (m == hash_')

-- | Newtype to allow the creation of a 'Checkable' typeclass for 
--   all 'ByteArrayAccess' without recurring to UndecidableInstances.
newtype Payload bs =  Payload bs

instance ByteArrayAccess bs => Checkable (Payload bs) where
  checkPayload hash_ (Payload p) = let hash' = convertString hash_ in do
    m <- toWeakMultihash hash'
    wmh <- weakMultihash (convertString $ show $ _getAlgorithm m) p
    check hash' wmh

-------------------------------------------------------------------------------

allowedAlgos :: [(ByteString, HashAlgo)]
allowedAlgos  = [ ("sha1", S1 A.SHA1) 
                , ("sha256", S256 A.SHA256)
                , ("sha512", S512 A.SHA512)
                , ("sha3-512", S3_512 A.SHA3_512)
                , ("sha3-384", S3_384 A.SHA3_384)
                , ("sha3-256", S3_256 A.SHA3_256)
                , ("sha3-224", S3_224 A.SHA3_224)
                , ("blake2b-512", B2b A.Blake2b_512)
                , ("blake2s-256", B2s A.Blake2s_256) ]

-- | Helper to multihash a 'ByteArrayAccess' (e.g. a 'BS.ByteString') using a 
--   supported hash algorithm. Uses 'Crypto.Hash.hash' for hashing.
weakMultihash :: ByteArrayAccess bs
                 => ByteString -> bs -> Either String WeakMultihashDigest
weakMultihash alg p = do
  alg' <- maybeToEither "Unknown algorithm" $ 
                        lookup (convertString alg) allowedAlgos
  let h = case alg' of
            S1 a     -> BA.convert $ hashWith a p
            S256 a   -> BA.convert $ hashWith a p
            S512 a   -> BA.convert $ hashWith a p
            S3_512 a -> BA.convert $ hashWith a p
            S3_384 a -> BA.convert $ hashWith a p
            S3_256 a -> BA.convert $ hashWith a p
            S3_224 a -> BA.convert $ hashWith a p
            B2b a    -> BA.convert $ hashWith a p
            B2s a    -> BA.convert $ hashWith a p
  return $ WeakMultihashDigest alg' (BA.length h) h

-- | Unsafe equivalent of 'weakMultihash'. Throws on creation errors.
weakMultihash' :: ByteArrayAccess bs => ByteString -> bs -> WeakMultihashDigest
weakMultihash' alg p = eitherToErr $ weakMultihash alg p

-- | Run the 'hashlazy' function but takes an explicit hash algorithm parameter
hashlazyWith :: A.HashAlgorithm alg => alg -> BL.ByteString -> Digest alg
hashlazyWith _ = hashlazy

-- | Helper to multihash a lazy 'BL.ByteString' using a supported hash algorithm.
--   Uses 'Crypto.Hash.hashlazy' for hashing.
weakMultihashlazy :: ByteString -> BL.ByteString -> Either String WeakMultihashDigest
weakMultihashlazy alg p = do
  alg' <- maybeToEither "Unknown algorithm" $ lookup alg allowedAlgos
  let h = case alg' of
            S1 a     -> BA.convert $ hashlazyWith a p
            S256 a   -> BA.convert $ hashlazyWith a p
            S512 a   -> BA.convert $ hashlazyWith a p
            S3_512 a -> BA.convert $ hashlazyWith a p
            S3_384 a -> BA.convert $ hashlazyWith a p
            S3_256 a -> BA.convert $ hashlazyWith a p
            S3_224 a -> BA.convert $ hashlazyWith a p
            B2b a    -> BA.convert $ hashlazyWith a p
            B2s a    -> BA.convert $ hashlazyWith a p
  return $ WeakMultihashDigest alg' (BA.length h) h

-- | Unsafe equivalent of 'weakMultihashlazy'. Throws on creation errors.
weakMultihashlazy' :: ByteString -> BL.ByteString -> WeakMultihashDigest
weakMultihashlazy' alg p = eitherToErr $ weakMultihashlazy alg p

-- | Helper to multihash a 'ByteArrayAccess' using a supported hash algorithm. 
--   Uses 'Crypto.Hash.hash' for hashing and truncates the hash to the lenght 
--   specified (must be positive and not longer than the digest length).
truncatedWeakMultihash :: ByteArrayAccess bs 
                       => Int -> ByteString -> bs -> Either String WeakMultihashDigest
truncatedWeakMultihash len alg bs = do
    WeakMultihashDigest alg' len' digest <- weakMultihash alg bs
    if len <= 0 || len > len'
      then Left "invalid truncated multihash lenght"
      else Right $ WeakMultihashDigest alg' len (BA.take len digest)

-- | Unsafe helper to multihash a 'ByteArrayAccess' using a supported hash algorithm. 
--   Uses 'Crypto.Hash.hash' for hashing and truncates the hash to the lenght 
--   specified (must be positive and not longer than the digest length, otherwise
--   the function will throw an error).
truncatedWeakMultihash' :: ByteArrayAccess bs 
                       => Int -> ByteString -> bs -> WeakMultihashDigest
truncatedWeakMultihash' len alg bs = eitherToErr $ truncatedWeakMultihash len alg bs

-------------------------------------------------------------------------------

-- | Imports a multihash into a 'WeahMultihashDigest' if possible.
toWeakMultihash :: BS.ByteString  -> Either String WeakMultihashDigest
toWeakMultihash bs = do
    base <- getBase bs
    h <- convertFromBase base bs
    if badLength h 
      then 
        Left "Corrupted MultihasDigest: invalid length"
      else do
        [b, b'] <- return $ take 2 $ BA.unpack h
        case elemIndex b hashCodes of
          Just 0 -> d h b' $ S1 A.SHA1
          Just 1 -> d h b' $ S256 A.SHA256
          Just 2 -> d h b' $ S512 A.SHA512
          Just 3 -> d h b' $ S3_512 A.SHA3_512
          Just 4 -> d h b' $ S3_384 A.SHA3_384
          Just 5 -> d h b' $ S3_256 A.SHA3_256
          Just 6 -> d h b' $ S3_224 A.SHA3_224
          Just 7 -> d h b' $ B2b A.Blake2b_512
          Just 8 -> d h b' $ B2s A.Blake2s_256
          Just _ -> Left "This should be impossible"
          Nothing -> Left "Impossible to infer the appropriate hash from the header"
  where 
    d h b' alg = Right $ WeakMultihashDigest alg (fromIntegral b') 
                                             (BA.convert $ BA.drop 2 h)

-------------------------------------------------------------------------------

-- | Safely check the correctness of an encoded 'Encodable' against the 
--   corresponding data.
checkWeakMultihash :: (IsString s, ConvertibleStrings s BS.ByteString, ByteArrayAccess bs)
                  => s -> bs -> Either String Bool
checkWeakMultihash h p = checkPayload h (Payload p)

-- | Unsafe version of 'checkWeakMultihash'. 
--   Throws on encoding/decoding errors instead of returning an 'Either' type.
checkWeakMultihash' :: (IsString s, ConvertibleStrings s BS.ByteString, ByteArrayAccess bs)
                   => s -> bs -> Bool
checkWeakMultihash' h p = checkPayload' h (Payload p)
